# Address tramp zero-click local command executation issue reported
# in August 2026 for Emacs 30.2 onwards, and possibly earlier.
#
#https://gitweb.gentoo.org/proj/emacs-patches.git/tree/emacs/30.2/05_all_tramp.patch
#
# Fix local command execution vulnerability in tramp
# Original patch by Michael Albinus <michael.albinus@gmx.de>,
# backported from Emacs 31
#
# Includes also a fix for bug#81622
#
# Disallow & in tramp-user-regexp.
#
--- old/doc/misc/tramp.texi	2025-04-05 13:01:10.000000000 +0200
+++ new/doc/misc/tramp.texi	2026-09-17 15:01:29.052844935 +0200
@@ -3601,7 +3601,9 @@ brackets @file{@value{ipv6prefix}} and @
 By default, @value{tramp} will use the current local user name as the
 remote user name for log in to the remote host.  Specifying a
 different name using the proper syntax will override this default
-behavior: @file{@trampfn{method,user@@host,path/to/file}}.
+behavior: @file{@trampfn{method,user@@host,path/to/file}}.  The user
+name can also be taken from an environment variable, like
+@file{@trampfn{method,@env{$MY_REMOTE_USER}@@host,path/to/file}}.
 
 @file{@trampfn{ssh,daniel@@melancholia,.emacs}} is for file
 @file{.emacs} in @samp{daniel}'s home directory on the host,
--- old/lisp/net/tramp-cmds.el	2025-04-05 13:01:10.000000000 +0200
+++ new/lisp/net/tramp-cmds.el	2026-09-17 15:01:29.053458270 +0200
@@ -574,33 +574,35 @@ For details, see `tramp-rename-files'."
   "Convert FILENAME into a multi-hop file name with \"sudo\".
 An alternative method could be chosen with `tramp-file-name-with-method'."
   (setq filename (expand-file-name filename))
-  (if (tramp-tramp-file-p filename)
-      (with-parsed-tramp-file-name filename nil
-	(cond
-	 ;; Remote file with proper method.
-	 ((string-equal method tramp-file-name-with-method)
-	  filename)
-	 ;; Remote file on the local host.
-	 ((and
-	   (stringp tramp-local-host-regexp) (stringp host)
-	   (string-match-p tramp-local-host-regexp host))
-	  (tramp-make-tramp-file-name
-	   (make-tramp-file-name
-	    :method tramp-file-name-with-method :localname localname)))
-	 ;; Remote file with multi-hop capable method.
-	 ((tramp-multi-hop-p v)
-	  (tramp-make-tramp-file-name
-	   (make-tramp-file-name
-	    :method (tramp-find-method tramp-file-name-with-method nil host)
-	    :user (tramp-find-user tramp-file-name-with-method nil host)
-	    :host (tramp-find-host tramp-file-name-with-method nil host)
-	    :localname localname :hop (tramp-make-tramp-hop-name v))))
-	 ;; Other remote file.
-	 (t (tramp-user-error v "Multi-hop with `%s' not applicable" method))))
-    ;; Local file.
-    (tramp-make-tramp-file-name
-     (make-tramp-file-name
-      :method tramp-file-name-with-method :localname filename))))
+  (expand-file-name
+   (if (tramp-tramp-file-p filename)
+       (with-parsed-tramp-file-name filename nil
+	 (cond
+	  ;; Remote file with proper method.
+	  ((string-equal method tramp-file-name-with-method)
+	   filename)
+	  ;; Remote file on the local host.
+	  ((and
+	    (stringp tramp-local-host-regexp) (stringp host)
+	    (string-match-p tramp-local-host-regexp host))
+	   (tramp-make-tramp-file-name
+	    (make-tramp-file-name
+	     :method tramp-file-name-with-method :localname localname)))
+	  ;; Remote file with multi-hop capable method.
+	  ((tramp-multi-hop-p v)
+	   (tramp-make-tramp-file-name
+	    (make-tramp-file-name
+	     :method (tramp-find-method tramp-file-name-with-method nil host)
+	     :user (tramp-find-user tramp-file-name-with-method nil host)
+	     :host (tramp-find-host tramp-file-name-with-method nil host)
+	     :localname localname :hop (tramp-make-tramp-hop-name v))))
+	  ;; Other remote file.
+	  (t
+	    (tramp-user-error v "Multi-hop with `%s' not applicable" method))))
+     ;; Local file.
+     (tramp-make-tramp-file-name
+      (make-tramp-file-name
+       :method tramp-file-name-with-method :localname filename)))))
 
 ;;;###tramp-autoload
 (defun tramp-revert-buffer-with-sudo ()
--- old/lisp/net/tramp-gvfs.el	2025-04-05 13:01:10.000000000 +0200
+++ new/lisp/net/tramp-gvfs.el	2026-09-17 15:01:29.054590920 +0200
@@ -1241,7 +1241,7 @@ file names."
 	  (when (string-match
 		 (rx bos "/" (+ (not "/")) (group "/.." (? "/"))) localname)
 	    (setq localname (replace-match "/" t t localname 1)))
-	(when (string-match (rx bol "/.." (? "/")) localname)
+	(when (string-match (rx bos "/.." (? "/")) localname)
 	  (setq localname (replace-match "/" t t localname))))
       ;; There might be a double slash.  Remove this.
       (while (string-match "//" localname)
@@ -1335,8 +1335,8 @@ If FILE-SYSTEM is non-nil, return file s
   (with-parsed-tramp-file-name filename nil
     (setq localname (file-name-unquote localname))
     (if (or (and (string-match-p
-		  (rx bol (| "afp" (: "dav" (? "s")) "smb") eol) method)
-		 (string-match-p (rx bol (? "/") (+ (not "/")) eol) localname))
+		  (rx bos (| "afp" (: "dav" (? "s")) "smb") eos) method)
+		 (string-match-p (rx bos (? "/") (+ (not "/")) eos) localname))
 	    (string-equal localname "/"))
 	(tramp-gvfs-get-root-attributes filename)
       (assoc
@@ -1773,14 +1773,14 @@ ID-FORMAT valid values are `string' and
   "Retrieve file name from D-Bus OBJECT-PATH."
   (dbus-unescape-from-identifier
    (replace-regexp-in-string
-    (rx bol (* nonl) "/" (group (+ (not "/"))) eol) "\\1" object-path)))
+    (rx bos (* nonl) "/" (group (+ (not "/"))) eos) "\\1" object-path)))
 
 (defun tramp-gvfs-url-host (url)
   "Return the host name part of URL, a string.
 We cannot use `url-host', because `url-generic-parse-url' returns
 a downcased host name only."
   (and (stringp url)
-       (string-match (rx bol (+ alnum) "://" (group (+ (not (any "/:"))))) url)
+       (string-match (rx bos (+ alnum) "://" (group (+ (not (any "/:"))))) url)
        (match-string 1 url)))
 
 ;; This is used in GNU ELPA package tramp-locproc.el.
@@ -1934,7 +1934,7 @@ Their full names are \"org.gtk.vfs.Mount
 		   (cadr (assoc "ssl" (cadr mount-spec)))))
 	     (uri (tramp-gvfs-dbus-byte-array-to-string
 		   (cadr (assoc "uri" (cadr mount-spec))))))
-	(when (string-match (rx bol (group (| "afp" "smb"))) method)
+	(when (string-match (rx bos (group (| "afp" "smb"))) method)
 	  (setq method (match-string 1 method)))
 	(when (and (string-equal "dav" method) (string-equal "true" ssl))
 	  (setq method "davs"))
@@ -2034,7 +2034,7 @@ Their full names are \"org.gtk.vfs.Mount
 		      (or
 		       (cadr (assoc "share" (cadr mount-spec)))
 		       (cadr (assoc "volume" (cadr mount-spec)))))))
-	 (when (string-match (rx bol (group (| "afp" "smb"))) method)
+	 (when (string-match (rx bos (group (| "afp" "smb"))) method)
 	   (setq method (match-string 1 method)))
 	 (when (and (string-equal "dav" method) (string-equal "true" ssl))
 	   (setq method "davs"))
@@ -2067,7 +2067,7 @@ Their full names are \"org.gtk.vfs.Mount
 		(string-equal host (tramp-file-name-host vec))
 		(string-equal port (tramp-file-name-port vec))
 		(string-match-p
-		 (rx bol "/" (literal (or share "")))
+		 (rx bos "/" (literal (or share "")))
 		 (tramp-file-name-unquote-localname vec)))
 	   ;; Set mountpoint and location.
 	   (tramp-set-file-property vec "/" "fuse-mountpoint" fuse-mountpoint)
@@ -2093,7 +2093,7 @@ Their full names are \"org.gtk.vfs.Mount
 (defun tramp-gvfs-mount-spec-entry (key value)
   "Construct a mount-spec entry to be used in a mount_spec.
 It was \"a(say)\", but has changed to \"a{sv})\"."
-  (if (string-match-p (rx bol "(aya{sv})") tramp-gvfs-mountlocation-signature)
+  (if (string-match-p (rx bos "(aya{sv})") tramp-gvfs-mountlocation-signature)
       (list :dict-entry key
 	    (list :variant (tramp-gvfs-dbus-string-to-byte-array value)))
     (list :struct key (tramp-gvfs-dbus-string-to-byte-array value))))
@@ -2112,9 +2112,9 @@ It was \"a(say)\", but has changed to \"
 		   (tramp-media-device-port media) (tramp-file-name-port vec)))
 	 (localname (tramp-file-name-unquote-localname vec))
 	 (share (when (string-match
-		       (rx bol (? "/") (group (+ (not "/")))) localname)
+		       (rx bos (? "/") (group (+ (not "/")))) localname)
 		  (match-string 1 localname)))
-	 (ssl (if (string-match-p (rx bol (| "davs" "nextcloud")) method)
+	 (ssl (if (string-match-p (rx bos (| "davs" "nextcloud")) method)
 		  "true" "false"))
 	 (mount-spec
           `(:array
@@ -2123,7 +2123,7 @@ It was \"a(say)\", but has changed to \"
                 (list (tramp-gvfs-mount-spec-entry "type" "smb-share")
                       (tramp-gvfs-mount-spec-entry "server" host)
                       (tramp-gvfs-mount-spec-entry "share" share)))
-               ((string-match-p (rx bol (| "davs" "nextcloud")) method)
+               ((string-match-p (rx bos (| "davs" "nextcloud")) method)
                 (list (tramp-gvfs-mount-spec-entry "type" "dav")
                       (tramp-gvfs-mount-spec-entry "host" host)
                       (tramp-gvfs-mount-spec-entry "ssl" ssl)))
@@ -2137,7 +2137,7 @@ It was \"a(say)\", but has changed to \"
                ((string-equal "nextcloud" method)
                 (list (tramp-gvfs-mount-spec-entry "type" "owncloud")
                       (tramp-gvfs-mount-spec-entry "host" host)))
-               ((string-match-p (rx bol "http") method)
+               ((string-match-p (rx bos "http") method)
                 (list (tramp-gvfs-mount-spec-entry "type" "http")
                       (tramp-gvfs-mount-spec-entry
 		       "uri"
@@ -2154,8 +2154,8 @@ It was \"a(say)\", but has changed to \"
             ,@(when port
                 (list (tramp-gvfs-mount-spec-entry "port" port)))))
 	 (mount-pref
-          (if (and (string-match-p (rx bol "dav") method)
-                   (string-match (rx bol (? "/") (+ (not "/"))) localname))
+          (if (and (string-match-p (rx bos "dav") method)
+                   (string-match (rx bos (? "/") (+ (not "/"))) localname))
               (match-string 0 localname)
 	    (tramp-gvfs-get-remote-prefix vec))))
 
--- old/lisp/net/tramp-rclone.el	2025-04-05 13:01:10.000000000 +0200
+++ new/lisp/net/tramp-rclone.el	2026-09-17 15:01:29.055019020 +0200
@@ -191,7 +191,7 @@ arguments to pass to the OPERATION."
   (with-tramp-connection-property nil "rclone-device-names"
     (tramp-compat-seq-keep
      (lambda (line)
-       (when (string-match (rx bol (group (+ (not blank))) ":" eol) line)
+       (when (string-match (rx bos (group (+ (not blank))) ":" eos) line)
 	 `(nil ,(match-string 1 line))))
      (tramp-process-lines nil tramp-rclone-program "listremotes"))))
 
@@ -365,7 +365,7 @@ file names."
 	  (tramp-rclone-maybe-open-connection v)
 	  ;; TODO: This shall be handled by `expand-file-name'.
 	  (setq localname
-		(replace-regexp-in-string (rx bol ".") "" (or localname "")))
+		(replace-regexp-in-string (rx bos ".") "" (or localname "")))
 	  (format "%s%s" (tramp-fuse-mounted-p v) localname)))
     ;; It is a local file name.
     filename))
--- old/lisp/net/tramp-smb.el	2025-04-05 13:01:10.000000000 +0200
+++ new/lisp/net/tramp-smb.el	2026-09-17 15:01:29.055747240 +0200
@@ -1550,7 +1550,7 @@ VEC or USER, or if there is no home dire
   "Return the share name of LOCALNAME."
   (save-match-data
     (let ((localname (tramp-file-name-unquote-localname vec)))
-      (when (string-match (rx bol (? "/") (group (+ (not "/"))) "/") localname)
+      (when (string-match (rx bos (? "/") (group (+ (not "/"))) "/") localname)
 	(match-string 1 localname)))))
 
 (defun tramp-smb-get-localname (vec)
@@ -1561,7 +1561,7 @@ If VEC has no cifs capabilities, exchang
       (setq
        localname
        (if (string-match
-	    (rx bol (? "/") (+ (not "/")) (group "/" (* nonl))) localname)
+	    (rx bos (? "/") (+ (not "/")) (group "/" (* nonl))) localname)
 	   ;; There is a share, separated by "/".
 	   (if (not (tramp-smb-get-cifs-capabilities vec))
 	       (mapconcat
@@ -1569,7 +1569,7 @@ If VEC has no cifs capabilities, exchang
 		(match-string 1 localname) "")
 	     (match-string 1 localname))
 	 ;; There is just a share.
-	 (if (string-match (rx bol (? "/") (group (+ (not "/"))) eol) localname)
+	 (if (string-match (rx bos (? "/") (group (+ (not "/"))) eos) localname)
 	     (match-string 1 localname)
 	   "")))
 
--- old/lisp/net/tramp.el	2025-05-10 13:38:42.000000000 +0200
+++ new/lisp/net/tramp.el	2026-09-17 15:02:02.065485645 +0200
@@ -973,11 +973,11 @@ Used in `tramp-make-tramp-file-name'.")
 
 (defun tramp-build-prefix-regexp ()
   "Return `tramp-prefix-regexp'."
-  (rx bol (literal (tramp-build-prefix-format))))
+  (rx bos (literal (tramp-build-prefix-format))))
 
 (defvar tramp-prefix-regexp nil ; Initialized when defining `tramp-syntax'!
   "Regexp matching the very beginning of Tramp file names.
-Should always start with \"^\".  Derived from `tramp-prefix-format'.")
+Should always start with \"\\\\=`\".  Derived from `tramp-prefix-format'.")
 
 (defconst tramp-method-regexp-alist
   `((default . ,(rx (| (literal tramp-default-method-marker) (>= 2 alnum))))
@@ -1019,7 +1019,10 @@ Used in `tramp-make-tramp-file-name'.")
   "Regexp matching delimiter between method and user or host names.
 Derived from `tramp-postfix-method-format'.")
 
-(defconst tramp-user-regexp (rx (+ (not (any "/:|" blank))))
+(defconst tramp-user-regexp
+  (rx (| (+ (not (any "/\\^$?*:;|[]{}()<>`'\"\&" blank)))
+	 ;; Environment variable.
+	 (: "$" (+ (any "_" alnum)))))
   "Regexp matching user names.")
 
 (defconst tramp-prefix-domain-format "%"
@@ -1801,6 +1804,8 @@ default values are used."
 	    (hop       (match-string (nth 5 tramp-file-name-structure) name))
 	    domain port v)
 	(when user
+	  (while (string-match (rx bos "$" (group (+ (any "_" alnum))) eos) user)
+	    (setq user (getenv (match-string 1 user))))
 	  (when (string-match tramp-user-with-domain-regexp user)
 	    (setq domain (match-string 2 user)
 		  user (match-string 1 user))))
@@ -2491,9 +2496,9 @@ Fall back to normal file name handler if
 	;; `file-remote-p' is called for everything, even for symbolic
 	;; links which look remote.  We don't want to get an error.
 	(non-essential (or non-essential (eq operation 'file-remote-p))))
+    (setq filename (tramp-replace-environment-variables filename))
     (if (tramp-tramp-file-p filename)
 	(save-match-data
-          (setq filename (tramp-replace-environment-variables filename))
           (with-parsed-tramp-file-name filename nil
             (let ((current-connection tramp-current-connection)
 		  (foreign
@@ -6537,6 +6542,9 @@ to cache the result.  Return the modifie
 			       (caar attr))
 			      (decode-coding-string
 			       (match-string 1 (caar attr)) 'utf-8))))
+	       ;; Quote remote-like symlink.
+	       (when (and (stringp (car attr)) (tramp-tramp-file-p (car attr)))
+		 (setcar attr (file-name-quote (car attr) 'top)))
 	       ;; Set file's gid change bit.
 	       (setcar
 		(nthcdr 9 attr)
--- old/test/lisp/net/tramp-tests.el	2025-05-17 12:45:04.000000000 +0200
+++ new/test/lisp/net/tramp-tests.el	2026-09-17 15:01:29.059878735 +0200
@@ -325,6 +325,7 @@ is greater than 10.
 	  (should (tramp-tramp-file-p "/method:user@:"))
 	  (should (tramp-tramp-file-p "/method:user@host:"))
 	  (should (tramp-tramp-file-p "/method:user@email@host:"))
+	  (should (tramp-tramp-file-p "/method:$USER@host:"))
 
 	  ;; Using a port.
 	  (should (tramp-tramp-file-p "/method:host#1234:"))
@@ -421,6 +422,7 @@ is greater than 10.
 	  (should (tramp-tramp-file-p "/user@:"))
 	  (should (tramp-tramp-file-p "/user@host:"))
 	  (should (tramp-tramp-file-p "/user@email@host:"))
+	  (should (tramp-tramp-file-p "/$USER@host:"))
 
 	  ;; Using a port.
 	  (should (tramp-tramp-file-p "/host#1234:"))
@@ -483,6 +485,7 @@ is greater than 10.
 	  (should (tramp-tramp-file-p "/[method/user@]"))
 	  (should (tramp-tramp-file-p "/[method/user@host]"))
 	  (should (tramp-tramp-file-p "/[method/user@email@host]"))
+	  (should (tramp-tramp-file-p "/[method/$USER@host]"))
 
 	  ;; Using a port.
 	  (should (tramp-tramp-file-p "/[method/host#1234]"))
@@ -671,6 +674,28 @@ is greater than 10.
 	  (should (string-equal
 		   (file-remote-p "/method:user@email@host:" 'hop) nil))
 
+	  ;; Expand environment variable.  It can be cascaded.
+	  (with-environment-variables
+	      (("REMOTE_USER" "$REMOTE_USER1") ("REMOTE_USER1" "remote-user"))
+	    (should (string-equal
+		     (file-remote-p "/method:$REMOTE_USER@host:")
+		     (format "/%s:%s@%s:" "method" "remote-user" "host")))
+	    (should
+	     (string-equal
+	      (file-remote-p "/method:$REMOTE_USER@host:" 'method) "method"))
+	    (should
+	     (string-equal
+	      (file-remote-p "/method:$REMOTE_USER@host:" 'user) "remote-user"))
+	    (should
+	     (string-equal
+	      (file-remote-p "/method:$REMOTE_USER@host:" 'host) "host"))
+	    (should
+	     (string-equal
+	      (file-remote-p "/method:$REMOTE_USER@host:" 'localname) ""))
+	    (should
+	     (string-equal
+	      (file-remote-p "/method:$REMOTE_USER@host:" 'hop) nil)))
+
 	  ;; Expand `tramp-default-method' and `tramp-default-user'.
 	  (should
 	   (string-equal
@@ -1151,6 +1176,28 @@ is greater than 10.
 	  (should (string-equal
 		   (file-remote-p "/user@email@host:" 'hop) nil))
 
+	  ;; Expand environment variable.  It can be cascaded.
+	  (with-environment-variables
+	      (("REMOTE_USER" "$REMOTE_USER1") ("REMOTE_USER1" "remote-user"))
+	    (should (string-equal
+		     (file-remote-p "/$REMOTE_USER@host:")
+		     (format "/%s@%s:" "remote-user" "host")))
+	    (should
+	     (string-equal
+	      (file-remote-p "/$REMOTE_USER@host:" 'method) "default-method"))
+	    (should
+	     (string-equal
+	      (file-remote-p "/$REMOTE_USER@host:" 'user) "remote-user"))
+	    (should
+	     (string-equal
+	      (file-remote-p "/$REMOTE_USER@host:" 'host) "host"))
+	    (should
+	     (string-equal
+	      (file-remote-p "/$REMOTE_USER@host:" 'localname) ""))
+	    (should
+	     (string-equal
+	      (file-remote-p "/$REMOTE_USER@host:" 'hop) nil)))
+
 	  ;; Expand `tramp-default-method' and `tramp-default-user'.
 	  (should (string-equal
 		   (file-remote-p "/host#1234:")
@@ -1617,6 +1664,28 @@ is greater than 10.
 	  (should (string-equal
 		   (file-remote-p "/[method/user@email@host]" 'hop) nil))
 
+	  ;; Expand environment variable.  It can be cascaded.
+	  (with-environment-variables
+	      (("REMOTE_USER" "$REMOTE_USER1") ("REMOTE_USER1" "remote-user"))
+	    (should (string-equal
+		     (file-remote-p "/[method/$REMOTE_USER@host]")
+		     (format "/[%s/%s@%s]" "method" "remote-user" "host")))
+	    (should
+	     (string-equal
+	      (file-remote-p "/[method/$REMOTE_USER@host]" 'method) "method"))
+	    (should
+	     (string-equal
+	      (file-remote-p "/[method/$REMOTE_USER@host]" 'user) "remote-user"))
+	    (should
+	     (string-equal
+	      (file-remote-p "/[method/$REMOTE_USER@host]" 'host) "host"))
+	    (should
+	     (string-equal
+	      (file-remote-p "/[method/$REMOTE_USER@host]" 'localname) ""))
+	    (should
+	     (string-equal
+	      (file-remote-p "/[method/$REMOTE_USER@host]" 'hop) nil)))
+
 	  ;; Expand `tramp-default-method' and `tramp-default-user'.
 	  (should (string-equal
 		   (file-remote-p "/[/host#1234]")
